### DISCUSSION TITLE
Publishing to the confluence

### DISCUSSION THREAD

**Discussion by z0lberg** [docToolchain/docToolchain#761] [Q&A]
Is there is a way to publish to the confluence with bearer token?

**Comment by rdmueller** [comment]
yes. let me check...

so, it depends what you mean by bearer token. I guess it is the token you can create within the confluence on demand cloud version.

As far as I know, you create the bearer token for a specific user. So try to set `confluenceUser` to the user-id and `confluencePass` to the token.

Is this what you need?

**Comment by z0lberg** [comment]
Hi! Looks like this is not working.
I am trying to test different combinations with a postman:
this is a working one -
![Screenshot 2022-01-25 at 15 18 39](https://user-images.githubusercontent.com/10826294/150994875-27a02b78-7982-40bf-9042-e6d466ceda6a.png)
![Screenshot 2022-01-25 at 15 21 05](https://user-images.githubusercontent.com/10826294/150994911-3dd8f716-2d82-4391-90e9-a10c167f3fe7.png)

and this was not - 
![Screenshot 2022-
![Screenshot 2022-01-25 at 15 22 33](https://user-images.githubusercontent.com/10826294/150995171-0f24540f-b11c-47a2-bf84-ecee3461af7a.png)
01-25 at 15 22 05](https://user-images.githubusercontent.com/10826294/150995088-85bc6b86-2503-472c-8c4c-505204286ec7.png)

Looks like docToolchain supports only Authorization: Basic from the asciidoc2confluence.groovy.

**Reply by rdmueller** [reply]
ok. so let's fix this...

the authorization header is distributed over the file:

https://github.com/docToolchain/docToolchain/blob/ng/scripts/asciidoc2confluence.groovy#L135

but that is the place where the logic has to be changed. 

* Do you use the locally installed version or the docker one?
* Where can I get such an API-Token for my confluence instance (I remember that I once used a user token which worked with the baisc auth machanism, but yours seems to be different).

**Comment by z0lberg** [comment]
I am using the docker one. 
You can try to generate the token in user profile -> settings -> personal access token
![Screenshot 2022-01-25 at 15 50 09](https://user-images.githubusercontent.com/10826294/150999875-6de07b30-dda2-47f3-9fa5-829020e0ddd9.png)
but it can be different depending on the confluence version etc.
![Screenshot 2022-01-25 at 15 54 37](https://user-images.githubusercontent.com/10826294/151000316-4c3a92df-787b-4e2f-b7cd-23c708874260.png)

**Reply by rdmueller** [reply]
hm. I am quite sure that with this token, the basic auth worked for me. But I have to verify.

regarding docker or local - with the local install, it is easier to modify the code. That's why I asked.

Can you please write a short issue ticket and reference this discussion?

**Reply by z0lberg** [reply]
Sure, done.

**Comment by z0lberg** [comment]
Hi! Can I try to help you?

**Reply by rdmueller** [reply]
yes, this would be helpful.

you find some information about how to set up a dev environment here http://doctoolchain.org/docToolchain/v2.0.x/025_development/010_setup_dev_env.html

I just scanned it and I guess it needs some clarification (it woudl also help to update these docs accordingly :-):

when you do a local install via

    ./dtcw local [someCommand]

like

    ./dtcw local tasks
    
docToolchain will be installed to you home folder `$HOME/.doctoolchain`. Basically, this is just a copy the github repository (minus the docs). So this lets you directly work on the code and test it. You can even clone the repository to the given location.
I like to create a fodler with a version like `2.0.3-dev` - this lets me easily switch between released versions and my working copy.

Does this help to get a start?

**Reply by z0lberg** [reply]
Yes thanks!

**Comment by z0lberg** [comment]
Could you please review it?
https://github.com/z0lberg/docToolchain/tree/feature/762-add-bearer-authentication

**Reply by rdmueller** [reply]
hey, great! Could you please open a pull request - it will be easier to review it this way..

**Reply by z0lberg** [reply]
Hi! Sure here is it: https://github.com/docToolchain/docToolchain/pull/785

**Comment by Strik0r** [comment]
Hey z0lberg, 

i'm trying exactly the same as you discriped.

Im using the wrapper, witch docker and the new Version 2.0.4.  

I tried publishToConfluence with the command: 
`.\dtcw publishToConfluence -PconfluenceUser=user -PconfluencePass=APIToken`
And I get the exeption:
```
something went wrong - got an http response code 401:
Reason : AUTHENTICATED_FAILEDBeschreibung The request has not been applied because it lacks valid authentication credentials for the target resource.Webserver
please check your confluence credentials in config file or passed parameters
> Task :publishToConfluence FAILED
FAILURE: Build failed with an exception.
* Where:
Script '/home/dtcuser/docToolchain/scripts/publishToConfluence.gradle' line: 26
* What went wrong:
Execution failed for task ':publishToConfluence'.
> java.lang.Exception: missing authentication credentials

```

If i call: 
`.\dtcw publishToConfluence`
With the config:
`credentials = "user:APIToken".bytes.encodeBase64().toString()`
I get exactly the same error message


If I call it with my password instead it works. Do I have to call it in another way if I use the API Token or what am I doing wrong?

**Reply by z0lberg** [reply]
Hi! You have to put bearerToken to your config.groovy
![image](https://user-images.githubusercontent.com/10826294/158799270-b9c30a96-6549-4174-b0f7-ac3faefc5ea5.png)

**Reply by Strik0r** [reply]
Hey 

with the bearerToken set in the docToolchainConfig.groovy  file it worked. Thank you very much! 
Is there a way to pass the confluence Token while calling the command? Something like: 
`.\dtcw publishToConfluence -bearerToken=token`

I haven't found any documentation about the bearer Token . Is there a another documentation besides the offical documentation with additional infos? 

It looks like you embeddet the ConfluenceToken as an environment variable. Where can i set this Variable?
I tried to read the ConfluenceToken out of an File like it was given in the example with the credentials 
 http://doctoolchain.org/docToolchain/v2.0.x/015_tasks/03_task_publishToConfluence.html but it always throws the error "no such file or directory"

For example: My confluence Token is stored in the File 
`C:\Users\strik0r\confluence\bearerToken` 
and if I put it in the config like:
`bearerToken = "${new File("C:\\Users\\strik0r\\confluence\\bearerToken").text}"`
I get the error message. In addition I tried it with only one "\" in the Path, with "/", with an relative Path in the project. 

I have no idear left. Maybe you can help me there too, or i can also include it via environment variable. 
In any case thank you really much for your help!

My goal is to write a github action, witch automaticly publishes the docs to confluence, when something gets pushed to the master branch. Therefore I need a Way to inject the bearer Token.

**Reply by z0lberg** [reply]
Hi! I am using an environment variable. Usually, you can find how to use it in your CI CD documentation. I guess this link should be helpful for you: https://docs.github.com/en/actions/learn-github-actions/environment-variables

**Reply by simonmysun** [reply]
~Hi, I may have mistaken something but I see nowhere mentioned bearerToken in `03_task_publishToConfluence.adoc`[^03_task_publishToConfluence.adoc] and nowhere refered to APIToken in `asciidoc2confluence.groovy`[^asciidoc2confluence.groovy]. Additionally, I see that in the document of enterprise version[^confluence_enterprice_doc_personal_token], basic authorization should be base64 encoded `username:password`, while in the cloud version document[^confluence_cloud_basic_auth] it's base64 encoded `your_email@domain.com:your_user_api_token`. This is very confusing to me. I have confirmed that both of them are correct and don't work the other way around. Is it because the APIs have changed and any of the document is outdated? If the outdated documents turn out to be docToolchain's, shall I try to update?~

[^03_task_publishToConfluence.adoc]: https://github.com/docToolchain/docToolchain/blob/fb8becb8dcc92e67bd01c5d7fb12ba35c8ac0193/src/docs/015_tasks/03_task_publishToConfluence.adoc "docToolchain/03_task_publishToConfluence.adoc at fb8becb8dcc92e67bd01c5d7fb12ba35c8ac0193 · docToolchain/docToolchain"
[^asciidoc2confluence.groovy]: https://github.com/docToolchain/docToolchain/blob/fb8becb8dcc92e67bd01c5d7fb12ba35c8ac0193/scripts/asciidoc2confluence.groovy "docToolchain/asciidoc2confluence.groovy at fb8becb8dcc92e67bd01c5d7fb12ba35c8ac0193 · docToolchain/docToolchain"
[^confluence_enterprice_doc_personal_token]: https://web.archive.org/web/20221229103704/https://confluence.atlassian.com/enterprise/using-personal-access-tokens-1026032365.html "Using Personal Access Tokens | Atlassian Support | Atlassian Documentation"
[^confluence_cloud_basic_auth]: https://web.archive.org/web/20220821111009/https://developer.atlassian.com/cloud/confluence/basic-auth-for-rest-apis/ "Basic auth for REST APIs"

Sorry, moved to https://github.com/docToolchain/docToolchain/issues/992
